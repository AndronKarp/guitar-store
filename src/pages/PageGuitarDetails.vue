<template>
  <b-row class="mb-3" cols-lg="2" cols="1">
    <b-col class="mb-3 mb-lg-0">
      <b-img
        fluid-grow
        :src="require(`../assets/${guitar.model}.png`)"
        :alt="guitar.model"
      />
    </b-col>
    <b-col>
      <h4>{{ guitar.brand }} {{ guitar.model }}</h4>
      <p>{{ guitar.description }}</p>
      <p class="text-success" v-if="isGuitarInStock(guitar)">In Stock</p>
      <p class="text-danger" v-else>Out Of Stock</p>
      <b-button
        :class="{ 'bg-info': isGuitarInStock(guitar) }"
        :disabled="!isGuitarInStock(guitar)"
        @click="addToCart(guitar)"
        >Add To Cart</b-button
      >
    </b-col>
  </b-row>
</template>

<script>
import { mapGetters } from "vuex";
import { guitarsRef } from "../configs/firebase";

export default {
  props: {
    slug: {
      type: String,
      required: true
    }
  },
  computed: {
    ...mapGetters(["guitars"]),
    guitar() {
      return this.guitars.find(guitar => guitar.slug === this.slug);
    },
    isGuitarInStock() {
      return guitar => guitar.quantity > 0;
    }
  },
  methods: {
    addToCart(guitar) {
      this.$store.dispatch("addToCart", guitar);
      guitarsRef
        .child(guitar.id)
        .update({ quantity: this.guitar.quantity - 1 });
    }
  }
};
</script>

<style></style>
