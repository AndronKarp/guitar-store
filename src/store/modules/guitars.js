export default {
  state: {
    guitars: [],
    areGuitarsFetched: false
  },
  getters: {
    guitars(state) {
      return state.guitars;
    },
    areGuitarsFetched(state) {
      return state.areGuitarsFetched;
    }
  },
  mutations: {
    pushToGuitars(state, guitar) {
      state.guitars.push(guitar);
    },
    setNewGuitarQuantity(state, { guitar, value }) {
      guitar.quantity = value;
    },
    updateAreGuitarsFetchedStatus(state, { status }) {
      state.areGuitarsFetched = status;
    }
  },
  actions: {
    addToGuitars(store, guitar) {
      store.commit("pushToGuitars", guitar);
    },
    updateGuitarQuantity(context, { guitar, value }) {
      context.commit("setNewGuitarQuantity", { guitar, value });
    },
    setAreGuitarsFetchedStatusToTrue(store) {
      store.commit("updateAreGuitarsFetchedStatus", { status: true });
    }
  }
};
