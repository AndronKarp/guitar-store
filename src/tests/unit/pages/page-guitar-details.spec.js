import { createLocalVue, shallowMount } from "@vue/test-utils";
import Vuex from "vuex";
import guitars from "@/store/modules/guitars";
import PageGuitarDetails from "@/pages/PageGuitarDetails";
import BootstrapVue from "bootstrap-vue";

const localVue = createLocalVue();

localVue.use(BootstrapVue);
localVue.use(Vuex);

describe("PageGuitarDetails.vue", () => {
  let wrapper;
  let store;
  let guitar;
  let state;
  let actions;

  beforeEach(() => {
    guitar = {
      brand: "Gibson",
      model: "SG",
      description: "Guitar",
      quantity: 1,
      slug: "sg"
    };
    state = { guitars: [guitar] };
    actions = {
      updateGuitarQuantity: jest.fn(),
      addToCart: jest.fn()
    };
    store = new Vuex.Store({
      state,
      getters: guitars.getters,
      actions
    });
    wrapper = shallowMount(PageGuitarDetails, {
      localVue,
      store,
      propsData: {
        slug: "sg"
      }
    });
  });

  test("renders guitar details", () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe("computed", () => {
    test("guitar returns guitars item finded by its slug", () => {
      expect(wrapper.vm.guitar).toEqual(guitar);
    });

    describe("isGuitarInStock", () => {
      test("returns true if quantity property of guitar is greater than zero", () => {
        expect(wrapper.vm.isGuitarInStock(guitar)).toBe(true);
      });

      test("returns false if quantiy property of guitar is equal to zero", () => {
        guitar.quantity = 0;
        expect(wrapper.vm.isGuitarInStock(guitar)).toBe(false);
      });
    });
  });

  describe("methods", () => {
    describe("addToCart", () => {
      test("calls addToCart and updateGuitarQuantity actions", () => {
        wrapper.vm.addToCart(guitar);
        expect(actions.addToCart).toHaveBeenCalled();
        expect(actions.updateGuitarQuantity).toHaveBeenCalled();
      });
    });
  });
});
