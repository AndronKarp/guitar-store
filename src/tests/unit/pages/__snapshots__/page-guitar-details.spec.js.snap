// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageGuitarDetails.vue renders guitar details 1`] = `
<b-row-stub tag="div" cols="1" colslg="2" class="mb-3">
  <b-col-stub tag="div" class="mb-3 mb-lg-0">
    <b-img-stub src="" alt="SG" fluidgrow="true" blankcolor="transparent"></b-img-stub>
  </b-col-stub>
  <b-col-stub tag="div">
    <h4>Gibson SG</h4>
    <p>Guitar</p>
    <p class="text-success">In Stock</p>
    <b-button-stub variant="secondary" type="button" tag="button" target="_self" class=" bg-info">Add To Cart</b-button-stub>
  </b-col-stub>
</b-row-stub>
`;
